File: vidyapath_app/lib/models/institution.dart
class InstitutionModel {
  final String id;
  final String name;
  final String? type;
  final String? affiliation;
  final Map<String, dynamic>? address;
  final Map<String, dynamic>? contact;
  final Map<String, dynamic>? facilities;
  final Map<String, dynamic>? stats;
  final Map<String, dynamic>? ratings;
  final bool isVerified;
  final List<String>? courses;
  final String? description;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  InstitutionModel({
    required this.id,
    required this.name,
    this.type,
    this.affiliation,
    this.address,
    this.contact,
    this.facilities,
    this.stats,
    this.ratings,
    this.isVerified = false,
    this.courses,
    this.description,
    this.createdAt,
    this.updatedAt,
  });

  factory InstitutionModel.fromJson(Map<String, dynamic> json) {
    return InstitutionModel(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      type: json['type'],
      affiliation: json['affiliation'],
      address: json['address'] is Map ? Map<String, dynamic>.from(json['address']) : null,
      contact: json['contact'] is Map ? Map<String, dynamic>.from(json['contact']) : null,
      facilities: json['facilities'] is Map ? Map<String, dynamic>.from(json['facilities']) : null,
      stats: json['stats'] is Map ? Map<String, dynamic>.from(json['stats']) : null,
      ratings: json['ratings'] is Map ? Map<String, dynamic>.from(json['ratings']) : null,
      isVerified: json['isVerified'] ?? false,
      courses: json['courses'] is List ? List<String>.from(json['courses']) : null,
      description: json['description'],
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt']) : null,
    );
  }

  String get city => address?['city'] ?? '';
  String get state => address?['state'] ?? '';
  String get district => address?['district'] ?? '';
  String get email => contact?['email'] ?? '';
  String get phone => contact?['phone'] ?? '';
  String get website => contact?['website'] ?? '';
  double get overallRating => (ratings?['overall'] as num?)?.toDouble() ?? 0.0;
  int get totalReviews => (ratings?['totalReviews'] as num?)?.toInt() ?? 0;
  bool get hasHostel => facilities?['hasHostel'] == true;
  bool get hasLibrary => facilities?['hasLibrary'] == true;
  bool get hasSports => facilities?['hasSports'] == true;
  bool get hasWifi => facilities?['hasWifi'] == true;
}

File: vidyapath/server/models/Institution.js
const mongoose = require('mongoose');

const institutionSchema = new mongoose.Schema({
  name: { type: String, required: true, trim: true },
  type: { type: String, enum: ['ITI', 'Diploma', 'College', 'University'], required: true },
  affiliation: { type: String, trim: true },
  address: {
    city: String, district: String, state: String, pincode: String,
    fullAddress: String,
  },
  location: {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: { type: [Number], index: '2dsphere' },
  },
  contact: {
    email: { type: String, default: '' },
    phone: { type: String, default: '' },
    website: { type: String, default: '' },
  },
  courses: [{ type: String, trim: true }],
  facilities: {
    hasHostel: { type: Boolean, default: false },
    hasLibrary: { type: Boolean, default: false },
    hasSports: { type: Boolean, default: false },
    hasWifi: { type: Boolean, default: false },
    hasCafeteria: { type: Boolean, default: false },
  },
  stats: {
    totalStudents: Number,
    placementRate: Number,
    avgFees: Number,
  },
  ratings: {
    overall: { type: Number, default: 0 },
    academics: { type: Number, default: 0 },
    infrastructure: { type: Number, default: 0 },
    faculty: { type: Number, default: 0 },
    placements: { type: Number, default: 0 },
    campus: { type: Number, default: 0 },
    valueForMoney: { type: Number, default: 0 },
    totalReviews: { type: Number, default: 0 },
  },
  customFields: { type: Map, of: mongoose.Schema.Types.Mixed, default: {} },
  isVerified: { type: Boolean, default: false },
  adminUsers: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  updatedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
}, { timestamps: true });

institutionSchema.index({ 'address.state': 1, 'address.district': 1 });

module.exports = mongoose.model('Institution', institutionSchema);
